src/libtiled/wangset.cpp

```diff
@@ -0,0 +1,166 @@
+/*
+ * wangset.cpp
+ * Copyright 2017, Benjamin Trotter <bdtrotte@ucsc.edu>
+ * This file is part of libtiled.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ *    1. Redistributions of source code must retain the above copyright notice,
+ *       this list of conditions and the following disclaimer.
+ *
+ *    2. Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in the
+ *       documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE CONTRIBUTORS ``AS IS'' AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO
+ * EVENT SHALL THE CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+ * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+ * OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+ * ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "wangset.h"
+
+#include <QStack>
+#include <QPoint>
+
+using namespace Tiled;
+
+WangSet::WangSet(Tileset *tileset,
+                 int edgeColors,
+                 int cornerColors,
+                 QString name,
+                 int imageTileId):
+    Object(Object::TerrainType), //for now, will add unique type soon
+    mTileSet(tileset),
+    mName(std::move(name)),
+    mImageTileId(imageTileId),
+    mEdgeColors(edgeColors),
+    mCornerColors(cornerColors)/*,
+    mWangIdToTile(QMultiHash<WangId, Tile*>()),
+    mTileIdToWangId(QHash<int, WangId>())*/
+{
+}
+
+void WangSet::addTile(Tile *tile, WangId wangId)
+{
+    Q_ASSERT(tile->tileset() == mTileSet);
+
+    for (int i = 0; i < 4; ++i) {
+        Q_ASSERT(wangId.edgeColor(i) <= mEdgeColors);
+    }
+
+    for (int i = 0; i < 4; ++i) {
+        Q_ASSERT(wangId.cornerColor(i) <= mCornerColors);
+    }
+
+    mWangIdToTile.insert(wangId, tile);
+    mTileIdToWangId.insert(tile->id(), wangId);
+}
+
+Tile *WangSet::findMatchingTile(WangId wangId) const
+{
+    auto potentials = findMatchingTiles(wangId);
+
+    if (potentials.length() > 0)
+        return potentials[qrand() % potentials.length()];
+    else
+        return NULL;
+}
+
+typedef struct WangWildCard
+{
+    int index, colorCount;
+} WangWildCard;
```

Comment:
This way of defining a struct was necessary in C times, to avoid needing to repeat the `struct` keyword whenever you would use `WangWildCard`. In C++, we can just write:
```c++
struct WangWildCard
{
    int index;
    int colorCount;
};
```